### PR TITLE
feat: add Rasa chatwidget page

### PR DESCRIPTION
**Pull request by fparga**
## Description
Add a new page `/rasa` that integrates a [chat widget to use with Rasa](https://github.com/botfront/rasa-webchat).

The page is only accessible via the URL for now.

image

Closes dialoguemd/covidflow#27, closes dialoguemd/covidflow#17, closes dialoguemd/covidflow#28,
closes dialoguemd/covidflow#29

## Checklist

- [X] PR title follows [Commit Convention](https://www.notion.so/godialogue/Commit-Convention-84fd9a4c149e48c998d760f1c9176df0) 
- [ ] PR title ends with a JIRA issue ID  
- [X] All relevant PR sections are populated, irrelevant ones are removed 
- [ ] [Draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests) is opened for WIP changes 
- [ ] PR review requested from groups not individuals 
- [X] PR link is posted to the corresponding Slack channel

### CODE REVIEW

File: .env.development [hunk 1]
```diff
@@ -1 +1,3 @@
 REACT_APP_FAQ_ENDPOINT="https://covidfaq.apps.dev.ca-central-1.dialoguecorp.com"
+REACT_APP_RASA_SOCKET_ENDPOINT="TBD"
```
Comment (vicrep) [inline .env.development:2]: these should be configured, no?

File: .env.production [hunk 2]
```diff
@@ -1 +1,3 @@
 REACT_APP_FAQ_ENDPOINT="https://covidfaq.dialoguecorp.com"
+REACT_APP_RASA_SOCKET_ENDPOINT="TBD"
+REACT_APP_RASA_SOCKET_PATH="/socket.io/"
```
Comment (vicrep) [inline .env.production:3]: missing newline at end of file

**Comment by vicrep** [comment]
also, deps need to be updated, there's a merge conflict with package.json and package-lock.json

File: .env.development (hunk 1 above)
Comment (fparga) [reply .env.development:2]: Yes, but the backend is not deployed yet. It should be soon, we should probably wait before merging this.

**Comment by fparga** [comment]
All done

**Comment by fparga** [comment]
@vicrep could you take another look a this?

File: src/components/rasa-chatbot.tsx [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+import { clamp } from 'lodash'
+
+import {
+  Widget,
+  toggle,
+  open,
+  close,
+  show,
+  hide,
+  toggleInputDisabled
```
Comment (vicrep) [inline src/components/rasa-chatbot.tsx:11]: it's kinda odd that this component exports global handlers for those kind of things, kinda breaks the functional principles of React. Does the component not have a more 'react' way of being controlled? (using states and props)

File: src/theme.ts [hunk 4]
```diff
@@ -7,6 +7,7 @@ const colors = {
   secondary: '#534E71',
   secondaryLight: '#7378A3',
   text: '#534E71',
+  lightText: '#FFFFFF',
```
Comment (vicrep) [inline src/theme.ts:10]:
```suggestion
  textLight: '#FFFFFF',
```

For consistency with other colour names.

File: package.json [hunk 5]
```diff
@@ -13,6 +13,7 @@
     "moment": "^2.24.0",
     "node-html-parser": "^1.2.12",
     "query-string": "^6.12.0",
+    "rasa-webchat": "git://github.com/nuecho/rasa-webchat#v0.8.5+nuecho.1",
```
[the remark this answers is not in the document]
Comment (fparga) [reply package.json:16]: No, not yet. I have submitted a PR last week on the original project. It has been merged but not released yet.

File: src/components/buttons.tsx [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import styled from 'styled-components/macro'
 
-export const CtaButton = styled.button`
+export const CtaButton = styled.a`
```
[the remark this answers is not in the document]
Comment (fparga) [reply src/components/buttons.tsx:3]: Good catch, leftover from before.

File: src/components/header.tsx [hunk 7]
```diff
@@ -14,12 +14,15 @@ const isLocalhost = window.location.hostname === 'localhost'
 const ENABLE_REGION_SWITCHING = isLocalhost || config.ENABLE_REGION_SWITCHING
 const { ENABLE_LOGO } = config
 
-interface Props {
+interface Props extends LogoContainerProps {
   title?: string
   showRegionPicker?: Boolean
 }
+interface LogoContainerProps {
+  hideLogoBackground?: Boolean
```
[the remark this answers is not in the document]
Comment (fparga) [reply src/components/header.tsx:22]: Right

File: src/components/rasa-chatbot.tsx [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+} from 'rasa-webchat'
+import styled from 'styled-components/macro'
+import { mobileBreakpoint } from 'theme'
```
[the remark this answers is not in the document]
Comment (fparga) [reply src/components/rasa-chatbot.tsx:14]: Ok, have you considered using this eslint plugin?
https://github.com/benmosher/eslint-plugin-import

The [`import/order` rule](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/order.md) can enforce this.

[the remark this answers is not in the document]
Comment (vicrep) [reply src/components/rasa-chatbot.tsx:14]: No, but I've used the equivalent import-order rule with tslint, I just didn't get to set either up on this repo yet.

Wondering, do you know if the eslint one is able to enforce vendor / app import separation? BC IIRC, the typescript one didn't (and only did ordering, which would organically sort relative imports at the end, but not with a line separation)

File: src/components/rasa-chatbot.tsx [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+    storage: 'local' | 'session'
+  }
+  profileAvatar: string
+  socketPath: string
+  socketUrl: string
+  subtitle: string
+  title: string
+}
+
+const onSocketEvent = {
```
[the remark this answers is not in the document]
Comment (fparga) [reply src/components/rasa-chatbot.tsx:29]: It sure can

File: src/components/rasa-chatbot.tsx [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+    }
+  }
+}
+
+const customMessageDelay = message => {
+  const delay = message.length * 15
+  return clamp(delay, 1500, 4000)
+}
+
+const WrappedWidget: React.FC<Props & Record<string, any>> = ({
```
[the remark this answers is not in the document]
Comment (fparga) [reply src/components/rasa-chatbot.tsx:47]: Removing

File: src/components/rasa-chatbot.tsx [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+  inputTextFieldHint,
+  params,
+  profileAvatar,
+  socketPath,
+  socketUrl,
+  subtitle,
+  title,
+  ...rest
+}) => {
+  useLayoutEffect(() => () => {
```
[the remark this answers is not in the document]
Comment (fparga) [reply src/components/rasa-chatbot.tsx:59]: No, I tried `useEffect` first but it doesn't work here. Adding a dependency though

File: src/components/rasa-chatbot.tsx [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+        socketUrl={socketUrl}
+        subtitle={subtitle}
+        title={title}
+        {...rest}
+      />
+    </div>
+  )
+}
+
+const RasaChatWidget = styled(WrappedWidget)`
```
[the remark this answers is not in the document]
Comment (fparga) [reply src/components/rasa-chatbot.tsx:84]: It's not how it's supposed to be done, but I can't directly style the component. The `className` is not rendered when passed to the `Widget` component that's why I have to wrap it in a `div`.

Is there another way to do this?

[the remark this answers is not in the document]
Comment (vicrep) [reply src/components/rasa-chatbot.tsx:84]: Unfortunately not if the component doesn't pass a className (tbh, it should).

In this case, I'd consider styling the top level `div` (you can call it `WidgetContainer` or wtv), as opposed to styling the `WrappedWidget` component :)

File: src/components/rasa-chatbot.tsx (hunk 3 above)
Comment (fparga) [reply src/components/rasa-chatbot.tsx:11]: Yeah I know it's weird, unfortunately that's all that is exposed :/

File: src/components/rasa-chatbot.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (fparga) [reply src/components/rasa-chatbot.tsx:15]: Yes it does if you set up the option `newlines-between` in your rule configuration.

File: src/components/rasa-chatbot.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (fparga) [reply src/components/rasa-chatbot.tsx]: Alright, done. It does look better!

**Comment by fparga** [comment]
@vicrep I think I have addressed all the requested changes. Let me know if there is something else that needs to be done before merging.

File: src/components/rasa-chatbot.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (vicrep) [reply src/components/rasa-chatbot.tsx:15]: Cool, I didn't know that! Thanks for the pointer, I'll add it to my laundry list :)

File: .prettierignore [hunk 13]
```diff
@@ -2,4 +2,5 @@
 *.jpeg
 *.png
 *.svg
+*.py
```
Comment (vicrep) [inline .prettierignore:5]: python in node projects make me sad 🙈

File: package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (vicrep) [reply package.json:16]: Great, thank you! As long as the branch it's pointing to is stable, then should be good :)

**Pull request merged**